chainer/functions/normalization/batch_normalization.py

```diff
@@ -439,6 +447,10 @@ def forward(self, inputs):
                                                 x, xp, expander,
                                                 self.mean, self.inv_std,
                                                 self.eps, self.var)
+        gx = gx.astype(x.dtype, copy=False)
+        ggamma = ggamma.astype(gamma.dtype, copy=False)
+        gbeta = gbeta.astype(gamma.dtype, copy=False)
```

Comment:
I think dtypes of  `gx` and `x` etc. should never be different. I mean it's backend impl's responsibility to return arrays with correct dtypes.